storage/shard.go

```diff
@@ -353,17 +359,49 @@ func (s *dbShard) OnRetrieveBlock(
 }
 
 func (s *dbShard) forEachShardEntry(entryFn dbShardEntryWorkFn) error {
+	return s.forEachShardEntryBatch(func(currEntries []*dbShardEntry) bool {
+		for _, entry := range currEntries {
+			if continueForEach := entryFn(entry); !continueForEach {
+				return false
+			}
+		}
+		return true
+	})
+}
+
+func (s *dbShard) forEachShardEntryBatch(entriesBatchFn dbShardEntryBatchWorkFn) error {
 	// NB(r): consider using a lockless list for ticking.
 	s.RLock()
 	elemsLen := s.list.Len()
 	batchSize := int(math.Ceil(shardIterateBatchPercent * float64(elemsLen)))
-	nextElem := s.list.Front()
 	s.RUnlock()
 
-	// TODO(xichen): pool or cache this.
-	currEntries := make([]*dbShardEntry, 0, batchSize)
-	for nextElem != nil {
+	decRefElem := func(e *list.Element) {
```

Comment:
nit: an allocation ever 1024 entries isn't much, but seems like you could just make this a method that you re-used since you're not creating any closures here

Reply:
re the allocation: I don't think it's ever going to grow much beyond 2K entries with our current prod workload (app/hist) so this won't ever be a bottleneck, but didn't want to drop the TODO without definitive proof. 

re the method: I considered doing that but liked having the method be self contained. I verified the function is marked safe for inlining so there's no performance cost: 

```
❯ go build -gcflags '-m' ./storage 2>&1 | grep 'shard.go' | sort | grep -e ':379'
storage/shard.go:379:16: (*dbShard).forEachShardEntryBatch func literal does not escape
storage/shard.go:379:16: can inline (*dbShard).forEachShardEntryBatch.func1
storage/shard.go:379:24: leaking param content: e
```